### PR TITLE
Feature: Support ERC1155 tokens

### PR DESCRIPTION
**Pull request by Femi-Ogunkola**

### CODE REVIEW

File: api/notd/block_processor.py [hunk 1]
```diff
@@ -59,12 +67,45 @@ async def get_transfers_in_block(self, blockNumber: int) -> List[RetrievedTokenT
...
+        registryAddress = event['address']
+        logging.debug(f'------------- {transactionHash} ------------')
+        if len(event['topics']) < 4:
+            logging.debug('Ignoring event with less than 4 topics')
+            return None
+        #print(f"trnasactionHash: {event['transactionHash'].hex()},toAddress: {event['topics'][3].hex()}")
+        operator = normalize_address(event['topics'][1].hex())
+        fromAddress = normalize_address(event['topics'][2].hex())
+        toAddress = normalize_address(event['topics'][3].hex())
+        tokenId =  int.from_bytes(bytes(event['topics'][1]), 'big')
```
Comment (Femi-Ogunkola) [inline api/notd/block_processor.py:96]: This is wrong but the tokenId isnt indexed so i am trying to find out how to get it

File: api/notd/block_processor.py [hunk 2]
```diff
@@ -59,12 +67,45 @@ async def get_transfers_in_block(self, blockNumber: int) -> List[RetrievedTokenT
         block = await self.ethClient.get_block(blockNumber)
         blockHash = block['hash'].hex()
         blockDate = datetime.datetime.fromtimestamp(block['timestamp'])
-        events = await self.ethClient.get_log_entries(startBlockNumber=blockNumber, endBlockNumber=blockNumber, topics=[self.erc721TansferEventSignatureHash])
-        logging.info(f'Found {len(events)} events in block #{blockNumber}')
-        allTokenTransfers = []
-        for eventsChunk in list_util.generate_chunks(events, 5):
-            allTokenTransfers += await asyncio.gather(*[self._process_event(event=dict(event), blockNumber=blockNumber, blockHash=blockHash, blockDate=blockDate) for event in eventsChunk])
-        return [tokenTransfer for tokenTransfer in allTokenTransfers if tokenTransfer]
+        #erc721events = await self.ethClient.get_log_entries(startBlockNumber=blockNumber, endBlockNumber=blockNumber, topics=[self.erc721TansferEventSignatureHash])
```
Comment (krishan711) [inline api/notd/block_processor.py:70]: why have you commented this out?

File: api/notd/block_processor.py [hunk 3]
```diff
@@ -59,12 +67,45 @@ async def get_transfers_in_block(self, blockNumber: int) -> List[RetrievedTokenT
...
+        #logging.info(f'Found {len(erc721events)} events in block #{blockNumber}')
+        erc1155events = await self.ethClient.get_log_entries(startBlockNumber=blockNumber, endBlockNumber=blockNumber, topics=[self.erc1155TansferEventSignatureHash])
+        print(erc1155events)
+        logging.info(f'Found {len(erc1155events)} erc1155events in block #{blockNumber}')
+
+        erc721TokenTransfers = []
+        erc1155TokenTransfers = []
+        #for erc721EventsChunk in list_util.generate_chunks(erc721events, 5):
```
Comment (krishan711) [inline api/notd/block_processor.py:78]: I would suggest you move all the below code into two functions like `_get_erc721_transfers(events=erc721events)` and `_get_erc1155_transfers(events=erc1155events)`

Comment (krishan711) [reply api/notd/block_processor.py:78]: (and then you can just comment out calling the function rather than having all this interwoven code below)

File: api/notd/block_processor.py (hunk 2 above)
Comment (krishan711) [reply api/notd/block_processor.py:70]: ah i see just for testing. i would suggest you change this to have a big erc721 block and then a big erc1155 block i.e. move all the commented lines up here so all the erc721 stuff from this function is together and then all the erc1155 stuff is together just below that.

File: api/notd/block_processor.py [hunk 4]
```diff
@@ -43,6 +43,14 @@ def __init__(self, ethClient: EthClientInterface):
         # self.contractFilter = self.ierc721Contract.events.Transfer.createFilter(fromBlock=6517190, toBlock=6517190, topics=[None, None, None, None])
         self.erc721TansferEventSignatureHash = Web3.keccak(text='Transfer(address,address,uint256)').hex()
 
+        with open('./contracts/IERC1155MetadataURI.json') as contractJsonFile:
+            contractJson = json.load(contractJsonFile)
+        self.ierc1155Contract = self.w3.eth.contract(abi=contractJson['abi'])
+        self.ierc721TransferEvent = self.ierc1155Contract.events.TransferSingle()
```
Comment (krishan711) [inline api/notd/block_processor.py:49]: we're gonna need to do both transfer single and transfer batch. could you paste her ewhat ths signatures are for both so we have a reference

Comment (krishan711) [inline api/notd/block_processor.py:49]: this variable has the wrong name

File: api/notd/model.py [hunk 5]
```diff
@@ -23,6 +23,23 @@ class RetrievedTokenTransfer:
...
     blockDate: datetime.datetime
 
+@dataclasses.dataclass(frozen=True, unsafe_hash=True)
+class ERC1155TokenTransfer:
+    transactionHash: str
+    registryAddress: str
+    operatorAddress: str
+    fromAddress: str
+    toAddress: str
+    tokenId: Optional[str]
```
Comment (krishan711) [inline api/notd/model.py:33]: these three shouldnt be optional

Comment (Femi-Ogunkola) [reply api/notd/model.py:33]: was testing something sorry

Comment (krishan711) [reply api/notd/model.py:33]: ah no worries, you can ignore my comments if its still wip 👍

File: api/notd/manager.py [hunk 6]
```diff
@@ -143,9 +143,15 @@ async def process_block(self, blockNumber: int, shouldForce: bool = False) -> No
                 return
         retrievedTokenTransfers = await self.blockProcessor.get_transfers_in_block(blockNumber=blockNumber)
         logging.info(f'Found {len(retrievedTokenTransfers)} token transfers in block #{blockNumber}')
-        await self._create_token_transfers(retrievedTokenTransfers=retrievedTokenTransfers)
-        await asyncio.gather(*[self.update_token_metadata_deferred(registryAddress=retrievedTokenTransfer.registryAddress, tokenId=retrievedTokenTransfer.tokenId) for retrievedTokenTransfer in retrievedTokenTransfers])
-        await asyncio.gather(*[self.update_collection_deferred(address=address) for address in set(retrievedTokenTransfer.registryAddress for retrievedTokenTransfer in retrievedTokenTransfers)])
+        if isinstance(retrievedTokenTransfers,list):
```
Comment (krishan711) [inline api/notd/manager.py:146]: This looks super wierd I would not have expected it. Can we work to undo these changes and make sure the response from get_transfers is always a list of token tranfers??

File: api/notd/block_processor.py [hunk 7]
```diff
@@ -96,5 +166,5 @@ async def _process_event(self, event: LogReceipt, blockNumber: int, blockHash: s
         value = ethTransaction['value']
         ethTransactionReceipt = await self._get_transaction_receipt(transactionHash=transactionHash)
         gasUsed = ethTransactionReceipt['gasUsed']
-        transaction = RetrievedTokenTransfer(transactionHash=transactionHash, registryAddress=registryAddress, fromAddress=fromAddress, toAddress=toAddress, tokenId=tokenId, value=value, gasLimit=gasLimit, gasPrice=gasPrice, gasUsed=gasUsed, blockNumber=blockNumber, blockHash=blockHash, blockDate=blockDate)
+        transaction = [RetrievedTokenTransfer(transactionHash=transactionHash, registryAddress=registryAddress, fromAddress=fromAddress, toAddress=toAddress, tokenId=tokenId, value=value, gasLimit=gasLimit, gasPrice=gasPrice, gasUsed=gasUsed, blockNumber=blockNumber, blockHash=blockHash, blockDate=blockDate)]
```
Comment (krishan711) [inline api/notd/block_processor.py:169]: Plural please for lists

Comment (krishan711) [reply api/notd/block_processor.py:169]: Actually it would probs be better to undo this line and put the array brackets on the line belpw

File: api/notd/block_processor.py [hunk 8]
```diff
@@ -59,12 +77,64 @@ async def get_transfers_in_block(self, blockNumber: int) -> List[RetrievedTokenT
...
+        logging.info(f'Found {len(erc1155events)} erc1155events in block #{blockNumber}')
+        for erc1155EventsChunk in list_util.generate_chunks(erc1155events, 5):
+            erc1155TokenTransfers += await asyncio.gather(*[self._process_erc1155_event(event=dict(event), blockNumber=blockNumber, blockHash=blockHash, blockDate=blockDate) for event in erc1155EventsChunk])
+
+        erc1155BatchTokenTransfers = []
+        erc1155Batchevents = await self.ethClient.get_log_entries(startBlockNumber=blockNumber, endBlockNumber=blockNumber, topics=[self.erc1155TansferBatchEventSignatureHash])
+        logging.info(f'Found {len(erc1155Batchevents)} erc1155Batchevents in block #{blockNumber}')
+        for erc1155BatchEventsChunk in list_util.generate_chunks(erc1155Batchevents, 5):
+            erc1155BatchTokenTransfers += await asyncio.gather(*[self._process_erc1155_event(event=dict(event), blockNumber=blockNumber, blockHash=blockHash, blockDate=blockDate) for event in erc1155BatchEventsChunk])
+
```
Comment (krishan711) [inline api/notd/block_processor.py:97]: I find it surprising that single and multiple can be done by the same function. Is this cos of how the arrays are structured?

File: api/notd/block_processor.py [hunk 9]
```diff
@@ -59,12 +77,64 @@ async def get_transfers_in_block(self, blockNumber: int) -> List[RetrievedTokenT
...
+        tokenTransferList = [tokenTransfer for tokenTransfer in erc1155BatchTokenTransfers if tokenTransfer] + [tokenTransfer for tokenTransfer in erc1155TokenTransfers if tokenTransfer] + [tokenTransfer for tokenTransfer in erc721TokenTransfers if tokenTransfer]
+        return  tokenTransferList
+
+    @staticmethod
+    def decode_transaction_data(data):
+        tokenIds = []
```
Comment (krishan711) [inline api/notd/block_processor.py:103]: I would suggest inlining this function - it's not really generic (very erc1155 specific) and is only used once.

File: api/notd/manager.py (hunk 6 above)
Comment (Femi-Ogunkola) [reply api/notd/manager.py:146]: `INFO:databases:Connected to database postgresql://postgres:********@127.0.0.1:5432/notddb
INFO:root:Found 1 erc1155Batchevents in block #13281282
INFO:root:Found 1 token transfers in block #13281282
INFO:root:Rolling back due to exception: 'list' object has no attribute 'transactionHash'
Traceback (most recent call last):
  File "/home/obafemi/Desktop/nftoftheday/api/local_testing_run.py", line 63, in <module>
    run(_anyio_backend='asyncio')
  File "/home/obafemi/.local/lib/python3.8/site-packages/asyncclick/core.py", line 1159, in __call__
    return anyio.run(self._main, main, args, kwargs, **({"backend":_anyio_backend} if _anyio_backend is not None else {}))
  File "/home/obafemi/.local/lib/python3.8/site-packages/anyio/_core/_eventloop.py", line 56, in run
    return asynclib.run(func, *args, **backend_options)  # type: ignore
  File "/home/obafemi/.local/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 230, in run
    return native_run(wrapper(), debug=debug)
  File "/usr/lib/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/obafemi/.local/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 225, in wrapper
    return await func(*args)
  File "/home/obafemi/.local/lib/python3.8/site-packages/asyncclick/core.py", line 1162, in _main
    return await main(*args, **kwargs)
  File "/home/obafemi/.local/lib/python3.8/site-packages/asyncclick/core.py", line 1083, in main
    rv = await self.invoke(ctx)
  File "/home/obafemi/.local/lib/python3.8/site-packages/asyncclick/core.py", line 1429, in invoke
    return await ctx.invoke(self.callback, **ctx.params)
  File "/home/obafemi/.local/lib/python3.8/site-packages/asyncclick/core.py", line 783, in invoke
    rv = await rv
  File "/home/obafemi/Desktop/nftoftheday/api/local_testing_run.py", line 53, in run
    await notdManager.process_block(blockNumber=blockNumber)
  File "/home/obafemi/Desktop/nftoftheday/api/notd/manager.py", line 152, in process_block
    await self._create_token_transfers(retrievedTokenTransfers=retrievedTokenTransfers)
  File "/home/obafemi/Desktop/nftoftheday/api/notd/manager.py", line 126, in _create_token_transfers
    StringFieldFilter(fieldName=TokenTransfersTable.c.transactionHash.key, eq=retrievedTokenTransfer.transactionHash),
AttributeError: 'list' object has no attribute 'transactionHash'`

Comment (Femi-Ogunkola) [reply api/notd/manager.py:146]: this is the error i got

Comment (krishan711) [reply api/notd/manager.py:146]: can you push the code that generates this error please

File: api/notd/block_processor.py [hunk 10]
```diff
@@ -59,12 +77,77 @@ async def get_transfers_in_block(self, blockNumber: int) -> List[RetrievedTokenT
...
+        amounts = data[4+lengthOfArray:4+lengthOfArray+lengthOfValue]
+        dataDict = {tokenIds[i]: amounts[i] for i in range(len(tokenIds))}
+        ethTransaction = await self._get_transaction(transactionHash=transactionHash)
+        gasLimit = ethTransaction['gas']
+        gasPrice = ethTransaction['gasPrice']
+        value = ethTransaction['value']
+        ethTransactionReceipt = await self._get_transaction_receipt(transactionHash=transactionHash)
+        gasUsed = ethTransactionReceipt['gasUsed']
+        transaction = [ERC1155TokenTransfer(transactionHash=transactionHash, operatorAddress=operatorAddress, registryAddress=registryAddress, fromAddress=fromAddress, toAddress=toAddress, tokenId=id, amount=amount ,value=value, gasLimit=gasLimit, gasPrice=gasPrice, gasUsed=gasUsed, blockNumber=blockNumber, blockHash=blockHash, blockDate=blockDate) for (id, amount) in dataDict.items()]
+        return transaction
```
Comment (krishan711) [inline api/notd/block_processor.py:150]:
```suggestion
        transactions = [ERC1155TokenTransfer(transactionHash=transactionHash, operatorAddress=operatorAddress, registryAddress=registryAddress, fromAddress=fromAddress, toAddress=toAddress, tokenId=id, amount=amount ,value=value, gasLimit=gasLimit, gasPrice=gasPrice, gasUsed=gasUsed, blockNumber=blockNumber, blockHash=blockHash, blockDate=blockDate) for (id, amount) in dataDict.items()]
        return transactions
```

File: api/notd/block_processor.py [hunk 11]
```diff
@@ -59,12 +77,77 @@ async def get_transfers_in_block(self, blockNumber: int) -> List[RetrievedTokenT
...
+        ethTransaction = await self._get_transaction(transactionHash=transactionHash)
+        gasLimit = ethTransaction['gas']
+        gasPrice = ethTransaction['gasPrice']
+        value = ethTransaction['value']
+        ethTransactionReceipt = await self._get_transaction_receipt(transactionHash=transactionHash)
+        gasUsed = ethTransactionReceipt['gasUsed']
+        transaction = ERC1155TokenTransfer(transactionHash=transactionHash, operatorAddress=operatorAddress, registryAddress=registryAddress, fromAddress=fromAddress, toAddress=toAddress, tokenId=tokenId, amount=amount ,value=value, gasLimit=gasLimit, gasPrice=gasPrice, gasUsed=gasUsed, blockNumber=blockNumber, blockHash=blockHash, blockDate=blockDate)
+        return transaction
+
+    async def _process_erc1155_batch_event(self, event: LogReceipt, blockNumber: int, blockHash: str, blockDate: datetime.datetime) -> Optional[RetrievedTokenTransfer]:
```
Comment (krishan711) [inline api/notd/block_processor.py:125]: return type is wrong here, make it a list and replace return None with return []

File: api/notd/block_processor.py [hunk 12]
```diff
@@ -59,12 +77,77 @@ async def get_transfers_in_block(self, blockNumber: int) -> List[RetrievedTokenT
...
+        erc1155BatchTokenTransfers = []
+        erc1155Batchevents = await self.ethClient.get_log_entries(startBlockNumber=blockNumber, endBlockNumber=blockNumber, topics=[self.erc1155TansferBatchEventSignatureHash])
+        logging.info(f'Found {len(erc1155Batchevents)} erc1155Batchevents in block #{blockNumber}')
+        for erc1155BatchEventsChunk in list_util.generate_chunks(erc1155Batchevents, 5):
+            erc1155BatchTokenTransfers += await asyncio.gather(*[self._process_erc1155_batch_event(event=dict(event), blockNumber=blockNumber, blockHash=blockHash, blockDate=blockDate) for event in erc1155BatchEventsChunk])
+
+        tokenTransferList = [tokenTransfer for tokenTransfer in erc1155BatchTokenTransfers for tokenTransfer in tokenTransfer] + [tokenTransfer for tokenTransfer in erc1155TokenTransfers if tokenTransfer] + [tokenTransfer for tokenTransfer in erc721TokenTransfers if tokenTransfer]
+        return  tokenTransferList
+
+    async def _process_erc1155_event(self, event: LogReceipt, blockNumber: int, blockHash: str, blockDate: datetime.datetime) -> Optional[RetrievedTokenTransfer]:
```
Comment (krishan711) [inline api/notd/block_processor.py:101]: make this return type a ist too and replace return None with return [] and change the last line to return [transaction]

Comment (krishan711) [reply api/notd/block_processor.py:101]: do the same with _process_event, and please rename that to _process_erc721_event

Comment (Femi-Ogunkola) [reply api/notd/block_processor.py:101]: I thought only batch returns a list.

File: api/notd/block_processor.py [hunk 13]
```diff
@@ -86,15 +164,16 @@ async def _process_event(self, event: LogReceipt, blockNumber: int, blockHash: s
...
         ethTransaction = await self._get_transaction(transactionHash=transactionHash)
+        operatorAddress = ethTransaction['from']
         gasLimit = ethTransaction['gas']
         gasPrice = ethTransaction['gasPrice']
         value = ethTransaction['value']
         ethTransactionReceipt = await self._get_transaction_receipt(transactionHash=transactionHash)
         gasUsed = ethTransactionReceipt['gasUsed']
-        transaction = RetrievedTokenTransfer(transactionHash=transactionHash, registryAddress=registryAddress, fromAddress=fromAddress, toAddress=toAddress, tokenId=tokenId, value=value, gasLimit=gasLimit, gasPrice=gasPrice, gasUsed=gasUsed, blockNumber=blockNumber, blockHash=blockHash, blockDate=blockDate)
-        return transaction
+        transactions = [RetrievedTokenTransfer(transactionHash=transactionHash, registryAddress=registryAddress, fromAddress=fromAddress, toAddress=toAddress, operatorAddress=operatorAddress, tokenId=tokenId, value=value, amount=1,gasLimit=gasLimit, gasPrice=gasPrice, gasUsed=gasUsed, blockNumber=blockNumber, blockHash=blockHash, blockDate=blockDate, tokenType='erc721')]
```
Comment (krishan711) [inline api/notd/block_processor.py:178]: spaces after commas when passing params pls

File: api/notd/block_processor.py [hunk 14]
```diff
@@ -59,14 +76,75 @@ async def get_transfers_in_block(self, blockNumber: int) -> List[RetrievedTokenT
...
+        data = [literal_eval(f'0x{elem}') for elem in data]
+        tokenId = data[0]
+        amount = data[1]
+        ethTransaction = await self._get_transaction(transactionHash=transactionHash)
+        gasLimit = ethTransaction['gas']
+        gasPrice = ethTransaction['gasPrice']
+        value = ethTransaction['value']
+        ethTransactionReceipt = await self._get_transaction_receipt(transactionHash=transactionHash)
+        gasUsed = ethTransactionReceipt['gasUsed']
+        transactions = [RetrievedTokenTransfer(transactionHash=transactionHash, registryAddress=registryAddress, fromAddress=fromAddress, toAddress=toAddress, operatorAddress=operatorAddress, tokenId=tokenId, amount=amount, value=value, gasLimit=gasLimit, gasPrice=gasPrice, gasUsed=gasUsed, blockNumber=blockNumber, blockHash=blockHash, blockDate=blockDate, tokenType='erc1155Single')]
```
Comment (krishan711) [inline api/notd/block_processor.py:117]: erc1155Single -> erc1155single

Comment (krishan711) [reply api/notd/block_processor.py:117]: no caps, none of the others have

File: api/notd/block_processor.py [hunk 15]
```diff
@@ -4,20 +4,21 @@
...
 from typing import Optional
+from ast import literal_eval
+import textwrap
 
 import async_lru
 from core.util import list_util
 from core.util.chain_util import normalize_address
 from core.web3.eth_client import EthClientInterface
-from web3 import Web3
+from web3.main import Web3
```
Comment (krishan711) [inline api/notd/block_processor.py:14]: can remove .main here i think

File: api/notd/block_processor.py [hunk 16]
```diff
@@ -59,14 +76,75 @@ async def get_transfers_in_block(self, blockNumber: int) -> List[RetrievedTokenT
...
+        logging.debug(f'------------- {transactionHash} ------------')
+        if len(event['topics']) < 4:
+            logging.debug('Ignoring event with less than 4 topics')
+            return []
+        operatorAddress = normalize_address(event['topics'][1].hex())
+        fromAddress = normalize_address(event['topics'][2].hex())
+        toAddress = normalize_address(event['topics'][3].hex())
+        data = event['data']
+        data = textwrap.wrap(data[2:], 64)
+        data = [literal_eval(f'0x{elem}') for elem in data]
```
Comment (krishan711) [inline api/notd/block_processor.py:108]: what is literal_eval doing here? kinda seems like it might not be needed

Comment (Femi-Ogunkola) [reply api/notd/block_processor.py:108]: using it to convert hex to -> int , the array count, token ids and values are stored has hexs

File: dbschema/1.tables.sql [hunk 17]
```diff
@@ -4,16 +4,19 @@ CREATE TABLE tbl_token_transfers (
     registry_address TEXT NOT NULL,
     from_address TEXT NOT NULL,
     to_address TEXT NOT NULL,
+    operator_address TEXT NOT NULL,
```
Comment (krishan711) [inline dbschema/1.tables.sql:7]: i think it might be worth adding an index for this?

File: dbschema/1.tables.sql [hunk 18]
```diff
@@ -4,16 +4,19 @@ CREATE TABLE tbl_token_transfers (
...
     gas_price NUMERIC(256, 0) NOT NULL,
     gas_used NUMERIC(256, 0) NOT NULL,
     block_number INTEGER NOT NULL,
     block_hash TEXT NOT NULL,
-    block_date TIMESTAMP WITHOUT TIME ZONE NOT NULL
+    block_date TIMESTAMP WITHOUT TIME ZONE NOT NULL,
+    token_type TEXT NOT NULL
 );
-CREATE UNIQUE INDEX tbl_token_transfers_transaction_hash_registry_address_token_id ON tbl_token_transfers (transaction_hash, registry_address, token_id);
+CREATE INDEX tbl_token_transfers_transaction_hash_registry_address_token_id ON tbl_token_transfers (transaction_hash, registry_address, token_id);
```
[the remark this answers is not in the document]
Comment (Femi-Ogunkola) [reply dbschema/1.tables.sql:19]: the solution to https://trello.com/c/xSSMGEcb its no longer a unique pairing

File: api/notd/block_processor.py (hunk 16 above)
Comment (krishan711) [reply api/notd/block_processor.py:108]: i dont understand if you remove the eval what breaks below?

File: dbschema/1.tables.sql (hunk 18 above)
[the remark this answers is not in the document]
Comment (Femi-Ogunkola) [reply dbschema/1.tables.sql:19]: okay no problem

File: api/notd/block_processor.py (hunk 16 above)
Comment (Femi-Ogunkola) [reply api/notd/block_processor.py:108]: length of Array becomes Hex
tokenIDs become hex and amount remain in hex

Comment (krishan711) [reply api/notd/block_processor.py:108]: ahh. is there not a better way to convert hex to string? maybe something in like web3 util. literal_eval is a super dangerous function so i would def prefer we avoid it if possible

File: api/tests/test_block_processor.py [hunk 19]
```diff
@@ -22,25 +22,31 @@ async def main():
...
+        RetrievedTokenTransfer(transactionHash='0x9e52523fc2de27af5e927d20a5b895024c46d7e9cebe758303c9768c3f756c19', registryAddress='0x495f947276749Ce646f68AC8c248420045cb7b5e', fromAddress='0xDc62e941fDDBDdDFc666B133E24E0C1aFae11847', toAddress='0xC4e245226537A0B1c5AF30079Eb0251504b42793', operatorAddress='0x750a32FAB9AE3B4ba3c15fF3d140b19aC92f25D8', tokenId='99683587586696680377209091673774721318238762396163454333212325529815842029569', amount=1, value=50000000000000000, gasLimit=228450, gasPrice=40000000000, gasUsed=218447, blockNumber=12839305, blockHash='0x7e36a9f127f3413595ef0f1c4165e4547d20ab566b66fc51a3cbe9eb048f3154', blockDate=datetime.datetime(2021, 7, 16, 18, 41, 41), tokenType='erc1155single')
     ]
-    assert result == expected
+    assert (result == expected)
```
Comment (krishan711) [inline api/tests/test_block_processor.py:48]: could you please add a block with a erc1155batch? not sure how you can find one but would be great to see

File: api/notd/model.py [hunk 20]
```diff
@@ -14,15 +14,17 @@ class RetrievedTokenTransfer:
     registryAddress: str
     fromAddress: str
     toAddress: str
+    operatorAddress: str
```
Comment (krishan711) [inline api/notd/model.py:17]: should be optional - may also have repurscussions elsewhere

**Review by krishan711 — APPROVED** [review]
